checker/nature/nature.go

```diff
@@ -1,261 +1,615 @@
 package nature
 
 import (
+	"fmt"
 	"reflect"
+	"time"
 
 	"github.com/expr-lang/expr/builtin"
 	"github.com/expr-lang/expr/internal/deref"
 )
 
 var (
-	unknown = Nature{}
+	intType      = reflect.TypeOf(0)
+	floatType    = reflect.TypeOf(float64(0))
+	arrayType    = reflect.TypeOf([]any{})
+	timeType     = reflect.TypeOf(time.Time{})
+	durationType = reflect.TypeOf(time.Duration(0))
+)
+
+type NatureCheck int
+
+const (
+	_ NatureCheck = iota
+	BoolCheck
+	StringCheck
+	IntegerCheck
+	NumberCheck
+	MapCheck
+	ArrayCheck
+	TimeCheck
+	DurationCheck
 )
 
 type Nature struct {
-	Type            reflect.Type      // Type of the value. If nil, then value is unknown.
-	Func            *builtin.Function // Used to pass function type from callee to CallNode.
-	ArrayOf         *Nature           // Elem nature of array type (usually Type is []any, but ArrayOf can be any nature).
-	PredicateOut    *Nature           // Out nature of predicate.
+	// The order of the fields matter, check alignment before making changes.
+
+	Type reflect.Type // Type of the value. If nil, then value is unknown.
+	Kind reflect.Kind // Kind of the value.
+
+	cache *Cache
+	*Optional
+	*FuncData
```

Comment:
Please also write the comment on Why? do we need separate *Optional and *FuncData so reader will understand it.